Spigot-Server-Patches/0718-Add-Channel-initialization-listeners.patch

```diff
@@ -0,0 +1,106 @@
++package io.papermc.paper.network;
++
++import net.kyori.adventure.key.Key;
++import org.checkerframework.checker.nullness.qual.NonNull;
++import org.checkerframework.checker.nullness.qual.Nullable;
++
++import java.util.Collections;
++import java.util.HashMap;
++import java.util.Map;
++
++/**
++ * Interal API to register channel initialization listeners.
++ * <p>
++ * This is not officially supported API and we make no guarantees to the existence or state of this class.
++ */
++public final class ChannelInitializeListenerHolder {
++
++    private static final Map<Key, ChannelInitializeListener> LISTENERS = new HashMap<>();
++
++    private ChannelInitializeListenerHolder() {
++    }
++
++    /**
++     * Registers whether an initialization listener is registered under the given key.
++     *
++     * @param key key
++     * @return whether an initialization listener is registered under the given key
++     */
++    public static boolean hasListener(@NonNull Key key) {
++        return LISTENERS.containsKey(key);
++    }
++
++    /**
++     * Registers a channel initialization listener called after ServerConnection is initialized.
++     *
++     * @param key      key
++     * @param listener initialization listeners
++     */
++    public static void addListener(@NonNull Key key, @NonNull ChannelInitializeListener listener) {
++        LISTENERS.put(key, listener);
++    }
++
++    /**
++     * Removes and returns an initialization listener registered by the given key if present.
++     *
++     * @param key key
++     * @return removed initialization listener if present
++     */
++    public static @Nullable ChannelInitializeListener removeListener(@NonNull Key key) {
++        return LISTENERS.remove(key);
++    }
++
++    /**
++     * Returns an immutable map of registered initialization listeners.
++     *
++     * @return immutable map of registered initialization listeners
++     */
++    public static @NonNull Map<Key, ChannelInitializeListener> getListeners() {
++        return Collections.unmodifiableMap(LISTENERS);
++    }
++}
+diff --git a/src/main/java/net/minecraft/server/network/ServerConnection.java b/src/main/java/net/minecraft/server/network/ServerConnection.java
+index dc362724ea0cc1b2f9d9ceffff483217b4356c40..028015fcc45b8f6d8d77c43088a51073eb7feb31 100644
+--- a/src/main/java/net/minecraft/server/network/ServerConnection.java
++++ b/src/main/java/net/minecraft/server/network/ServerConnection.java
+@@ -104,6 +104,7 @@ public class ServerConnection {
+                     pending.add((NetworkManager) object); // Paper
+                     channel.pipeline().addLast("packet_handler", (ChannelHandler) object);
+                     ((NetworkManager) object).setPacketListener(new HandshakeListener(ServerConnection.this.e, (NetworkManager) object));
++                    io.papermc.paper.network.ChannelInitializeListenerHolder.getListeners().values().forEach(listener -> listener.afterInitChannel(channel)); // Paper
```

Comment:
Maybe better to add a method `ChannelInitializeListenerHolder#afterChannelInit ` that if not empty directly iterate on the Map values instead of calling `getListeners` which create `unmodifiableMap` ?

Reply:
Yeah sounds good, done.